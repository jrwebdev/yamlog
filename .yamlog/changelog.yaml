0.1.4:
  metadata:
    date: '2018-04-07'
  fix:
    - module: publish
      details: Fix error committing changelog
0.1.3:
  metadata:
    date: '2017-10-28'
  feature:
    - module: log
      details: Allow module input question to be disabled
0.1.2:
  metadata:
    date: '2017-10-28'
  fix:
    - module: commit
      details: Fix multiple changes not being logged on commit
0.1.1:
  metadata:
    date: '2017-10-28'
  feature:
    - module: publish
      details: Check branch is master and up-to-date before publishing
0.1.0:
  metadata:
    date: '2017-10-26'
  breaking:
    - >-
      Move changelog.yaml to .yamlog/changelog.yaml, remove unreleased section
      from changelog and always log changes to unreleased directory
0.0.12:
  metadata:
    date: '2017-10-25'
  feature:
    - module: bump-version
      details: Output module to CHANGELOG.md
      logAnother: false
  fix:
    - module: commit
      details: Do not add logAnother to changelog
0.0.11:
  metadata:
    date: '2017-10-25'
  feature:
    - module: commit
      details: Log module from commit
      logAnother: false
0.0.10:
  metadata:
    date: '2017-10-25'
  feature:
    - details: Allow module for change to be logged
      module: log
0.0.9:
  metadata:
    date: '2017-10-25'
  fix:
    - Fix publish git push for travis ci
0.0.8:
  metadata:
    date: '2017-10-25'
  feature:
    - use-commit CLI command for checking that git commit isn't being used
  fix:
    - Fix commit not adding to changelog
0.0.7:
  metadata:
    date: '2017-10-23'
  feature:
    - Commit CLI command
0.0.6:
  metadata:
    date: '2017-10-22'
  feature:
    - Publish CLI script to bump version and publish changes to npm
0.0.5:
  metadata:
    date: '2017-10-22'
  feature:
    - changes CLI command to output changes to the command line
0.0.4:
  metadata:
    date: '2017-10-21'
  feature:
    - Generate CHANGELOG.md on version bump
