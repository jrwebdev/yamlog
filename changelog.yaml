0.0.9:
  metadata:
    date: '2017-10-25'
  fix:
    - Fix publish git push for travis ci
0.0.8:
  metadata:
    date: '2017-10-25'
  feature:
    - use-commit CLI command for checking that git commit isn't being used
  fix:
    - Fix commit not adding to changelog
0.0.7:
  metadata:
    date: '2017-10-23'
  feature:
    - Commit CLI command
0.0.6:
  metadata:
    date: '2017-10-22'
  feature:
    - Publish CLI script to bump version and publish changes to npm
0.0.5:
  metadata:
    date: '2017-10-22'
  feature:
    - changes CLI command to output changes to the command line
0.0.4:
  metadata:
    date: '2017-10-21'
  feature:
    - Generate CHANGELOG.md on version bump
