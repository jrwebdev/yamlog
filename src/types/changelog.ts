export type ChangeType = 'breaking' | 'feature' | 'fix';

export interface DetailedChange {
  details: string;
  module?: string;
}

export type Change = string | DetailedChange;

export interface ChangelogVersionMetadata {
  date: string;
}

type ChangelogVersionChanges = { [type in ChangeType]?: Change[] };

export interface ChangelogVersion extends ChangelogVersionChanges {
  metadata?: ChangelogVersionMetadata;
}

export interface Changelog {
  [version: string]: ChangelogVersion;
}
