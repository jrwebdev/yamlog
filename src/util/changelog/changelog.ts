import { compose } from 'ramda';
import {
  ChangeType,
  Change,
  DetailedChange,
  ChangelogVersion,
} from './changelog.types';

import * as unreleasedDir from './unreleased-dir';

const changeTypeKeys: ChangeType[] = ['breaking', 'feature', 'fix'];

const isDetailedChange = (change: any): change is DetailedChange =>
  typeof change === 'object';

const stripEmptyChangeTypes = (version: ChangelogVersion) =>
  changeTypeKeys.reduce(
    (acc, key) => {
      // TODO: Fix type
      if (acc[key] && !(acc as any)[key].length) {
        delete acc[key];
      }
      return acc;
    },
    { ...version }
  );

const stripEmptyChangeProperties = (change: Change): DetailedChange | Change =>
  isDetailedChange(change)
    ? Object.keys(change).reduce((acc, property) => {
        const value = (change as any)[property];
        if (value && value.trim()) {
          (acc as any)[property] = value;
        }
        return acc;
      }, {}) as DetailedChange
    : change;

const flattenDetails = (change: Change) =>
  typeof change !== 'string' &&
  Object.keys(change).length === 1 &&
  change.details
    ? change.details
    : change;

const processChange = compose(flattenDetails, stripEmptyChangeProperties);

export const addChange = (changeType: ChangeType, change: Change) => {
  const processedChange = processChange(change);
  unreleasedDir.addChange(changeType, processedChange);
};
