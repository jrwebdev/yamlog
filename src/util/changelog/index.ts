export * from './changelog';
export * from './changelog.types';
