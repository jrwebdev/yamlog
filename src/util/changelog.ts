import * as moment from 'moment';

import { ChangelogConfig, CurrentVersionLoader } from '../types/config';
import { Changelog, ChangeType, Change } from '../types/changelog';

import {
  addChange as addUnreleasedDirChange,
  read as readUnreleasedFiles,
  removeDir as deleteUnreleasedFiles,
} from './unreleased-dir';
import { getNextVersion } from './changelog-helpers';
import {
  addChange as addChangelogChange,
  read as readChangelog,
  write as writeChangelog,
} from './changelog-file';

const loadCurrentVersion = async (
  loaders: CurrentVersionLoader,
  releasedVersions: Changelog
) => {
  if (Array.isArray(loaders)) {
    let version;
    let i = 0;
    // TODO: Change to use recursion
    while (!version && loaders[i]) {
      version = await loaders[i](releasedVersions);
      i += 1;
    }
    return version;
  } else {
    return await loaders(releasedVersions);
  }
};

export const addChange = (
  changeType: ChangeType,
  change: Change,
  { unreleasedDir = '' } = {}
) =>
  unreleasedDir
    ? addUnreleasedDirChange(changeType, change, unreleasedDir)
    : addChangelogChange(changeType, change);

export const bumpVersion = async (
  currentVersionLoader: CurrentVersionLoader,
  config: ChangelogConfig = {}
) => {
  const {
    unreleased: changelogUnreleased,
    ...released,
  } = await readChangelog();

  // TODO: Combine?
  const unreleased = config.unreleasedDir
    ? await readUnreleasedFiles(config.unreleasedDir)
    : changelogUnreleased;

  if (unreleased && Object.keys(unreleased).length) {
    const currentVersion = await loadCurrentVersion(
      currentVersionLoader,
      released
    );

    const newVersion = getNextVersion(unreleased, currentVersion, config);
    if (newVersion) {
      await writeChangelog({
        [newVersion]: {
          metadata: {
            date: moment().format('YYYY-MM-DD'),
          },
          ...unreleased,
        },
        ...released,
      });

      if (config.unreleasedDir) {
        await deleteUnreleasedFiles(config.unreleasedDir);
      }

      return newVersion;
    }
  }

  return undefined;
};
