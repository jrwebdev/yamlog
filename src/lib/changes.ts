import { ChangelogConfig } from '../types/config';
import { Changelog, ChangelogVersion } from '../types/changelog';
import { VersionQuery } from '../types/version';

import * as changelog from '../util/changelog';
import * as yaml from '../util/yaml';
import markdownTransformer from '../util/transformers/markdown';

export interface ChangesOptions {
  transform?: 'yaml' | 'markdown' | 'json' | 'text';
  version?: VersionQuery;
}

// TODO: Move
const isChangelogVersion = (version: any): version is ChangelogVersion =>
  typeof version === 'object' &&
  (version.metadata || version.breaking || version.feature || version.fix);

// TODO: Move
const transformMarkdown = (versions: Changelog | ChangelogVersion) =>
  isChangelogVersion(versions)
    ? markdownTransformer(versions)
    : Object.keys(versions)
        .map(version => {
          const { metadata, ...changes } = versions[version];
          return markdownTransformer(changes, version, metadata);
        })
        .join('\n');

export default async (
  options: ChangesOptions = {},
  config: ChangelogConfig = {}
) => {
  const changes = await changelog.getChanges(options.version, {
    unreleasedDir: config.unreleasedDir || '',
  });

  switch (options.transform) {
    case 'yaml':
      return yaml.format(changes);
    case 'json':
      return JSON.stringify(changes, null, 2);
    case 'markdown':
      return transformMarkdown(changes);
    default:
      return changes;
  }
};
