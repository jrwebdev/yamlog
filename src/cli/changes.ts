import config from '../util/config';
import { default as changesLib } from '../lib/changes';

const run = async () => {
  const transform = 'yaml';
  const changes = await changesLib({ transform }, config);
  console.log(changes);
};

run();
