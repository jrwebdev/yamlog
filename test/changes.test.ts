import { existsSync, readFile } from 'fs-extra';
import { createMockProject } from './test-utils';

import { Changelog } from '../src/types/changelog';
import changes from '../src/lib/changes';

let mockProject: any;
let mockChangelog: Changelog;

beforeEach(() => {
  mockChangelog = {
    unreleased: {
      breaking: ['Breaking Change 1'],
      feature: ['Feature 1', 'Feature 2'],
      fix: ['Fix 1', 'Fix 2'],
    },
    '2.1.0': {
      metadata: { date: '2017-01-05' },
      feature: ['Feature 1', 'Feature 2'],
      fix: ['Fix 1'],
    },
    '2.0.1': {
      metadata: { date: '2017-01-04' },
      fix: ['Fix 1'],
    },
    '2.0.0': {
      metadata: { date: '2017-01-03' },
      breaking: ['Breaking Change 1'],
      feature: ['Feature 1'],
    },
    '1.0.0': {
      metadata: { date: '2017-01-02' },
      feature: ['Feature 1'],
    },
    '0.2.0': {
      metadata: { date: '2017-01-01' },
      feature: ['Feature 1'],
    },
    '0.1.0': {
      metadata: { date: '2017-01-01' },
      feature: ['Feature 1'],
    },
  };

  mockProject = createMockProject();
  mockProject.setup(mockChangelog);
  expect.hasAssertions();
});

afterEach(() => {
  mockProject.teardown();
});

const testSnapshot = (expected: any) => {
  // Needed so loading the snapshot files isn't mocked
  mockProject.teardown();
  expect(expected).toMatchSnapshot();
};

it('should return all changes for the project except unreleased', async () => {
  const c = await changes();
  testSnapshot(JSON.stringify(c, null, 2));
});

it('should return changes in JSON', async () => {
  const c = await changes({ format: 'json' });
  testSnapshot(c);
});

it('should return changes in Yaml', async () => {
  const c = await changes({ format: 'yaml' });
  testSnapshot(c);
});

it('should return changes in markdown', async () => {
  const c = await changes({ format: 'markdown' });
  testSnapshot(c);
});

// TODO
xit('should return changes in plain text', async () => {
  const c = await changes({ format: 'text' });
  testSnapshot(c);
});

it('should return a single version', async () => {
  const c = await changes({ format: 'yaml', version: '2.0.0' });
  testSnapshot(c);
});

it('should return a single version in markdown', async () => {
  const c = await changes({ format: 'markdown', version: '2.0.0' });
  testSnapshot(c);
});

it('should return unreleased changes', async () => {
  const c = await changes({ format: 'yaml', version: 'unreleased' });
  testSnapshot(c);
});

// TODO: Fix
it('should return unreleased changes in markdown', async () => {
  const c = await changes({ format: 'markdown', version: 'unreleased' });
  testSnapshot(c);
});

it('should return unreleased changes from change files', async () => {
  mockProject = createMockProject();
  mockProject.setup(
    {
      '0.1.0': {
        metadata: { date: '2017-01-01' },
        feature: ['Feature 1'],
      },
    },
    {
      breaking: ['Breaking 1', 'Breaking 2'],
      feature: ['Feature 1', 'Feature 2'],
      fix: ['Fix 1', 'Fix 2'],
    }
  );

  const c = await changes({ format: 'yaml', version: 'unreleased' });
  testSnapshot(c);
});

it('should return the latest version', async () => {
  const c = await changes({ format: 'yaml', version: 'latest' });
  testSnapshot(c);
});

it('should return a range of versions', async () => {
  const c = await changes({
    format: 'yaml',
    version: { from: '0.2.0', to: '2.0.0' },
  });
  testSnapshot(c);
});

it('should filter and sort versions correctly when returning a range', async () => {
  mockChangelog = {
    unreleased: {
      fix: ['Fix 1'],
    },
    '10.0.0': {
      metadata: { date: '2017-01-01' },
      fix: ['Fix 1'],
    },
    '2.0.0': {
      metadata: { date: '2017-01-01' },
      fix: ['Fix 1'],
    },
    '1.10.0': {
      metadata: { date: '2017-01-01' },
      fix: ['Fix 1'],
    },
    '1.2.0': {
      metadata: { date: '2017-01-01' },
      fix: ['Fix 1'],
    },
    '1.1.0': {
      metadata: { date: '2017-01-01' },
      fix: ['Fix 1'],
    },
    '1.1.100': {
      metadata: { date: '2017-01-01' },
      fix: ['Fix 1'],
    },
    '1.1.2': {
      metadata: { date: '2017-01-01' },
      fix: ['Fix 1'],
    },
    '1.1.1': {
      metadata: { date: '2017-01-01' },
      fix: ['Fix 1'],
    },
    '1.0.0': {
      metadata: { date: '2017-01-01' },
      fix: ['Fix 1'],
    },
    '0.1.0': {
      metadata: { date: '2017-01-01' },
      fix: ['Fix 1'],
    },
  };
  mockProject = createMockProject();
  mockProject.setup(mockChangelog);
  const c = await changes({
    version: { from: '1.0.0', to: '2.0.0' },
  });
  expect(Object.keys(c)).toEqual([
    '2.0.0',
    '1.10.0',
    '1.2.0',
    '1.1.100',
    '1.1.2',
    '1.1.1',
    '1.1.0',
    '1.0.0',
  ]);
});

it('should return a version from a specified version to the latest version', async () => {
  const c = await changes({ format: 'yaml', version: { from: '1.0.0' } });
  testSnapshot(c);
});
