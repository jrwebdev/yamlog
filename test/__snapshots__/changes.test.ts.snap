// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return a range of versions 1`] = `
2.0.0:
  metadata:
    date: '2017-01-03'
  breaking:
    - Breaking Change 1
  feature:
    - Feature 1
1.0.0:
  metadata:
    date: '2017-01-02'
  feature:
    - Feature 1
0.2.0:
  metadata:
    date: '2017-01-01'
  feature:
    - Feature 1

`;

exports[`should return a single version 1`] = `
2.0.0:
  metadata:
    date: '2017-01-03'
  breaking:
    - Breaking Change 1
  feature:
    - Feature 1

`;

exports[`should return a single version in markdown 1`] = `
## 2.0.0 (2017-01-03)
### BREAKING CHANGES
* Breaking Change 1
### Features
* Feature 1

`;

exports[`should return a version from a specified version to the latest version 1`] = `
2.1.0:
  metadata:
    date: '2017-01-05'
  feature:
    - Feature 1
    - Feature 2
  fix:
    - Fix 1
2.0.1:
  metadata:
    date: '2017-01-04'
  fix:
    - Fix 1
2.0.0:
  metadata:
    date: '2017-01-03'
  breaking:
    - Breaking Change 1
  feature:
    - Feature 1
1.0.0:
  metadata:
    date: '2017-01-02'
  feature:
    - Feature 1

`;

exports[`should return all changes for the project except unreleased 1`] = `
{
  "2.1.0": {
    "metadata": {
      "date": "2017-01-05"
    },
    "feature": [
      "Feature 1",
      "Feature 2"
    ],
    "fix": [
      "Fix 1"
    ]
  },
  "2.0.1": {
    "metadata": {
      "date": "2017-01-04"
    },
    "fix": [
      "Fix 1"
    ]
  },
  "2.0.0": {
    "metadata": {
      "date": "2017-01-03"
    },
    "breaking": [
      "Breaking Change 1"
    ],
    "feature": [
      "Feature 1"
    ]
  },
  "1.0.0": {
    "metadata": {
      "date": "2017-01-02"
    },
    "feature": [
      "Feature 1"
    ]
  },
  "0.2.0": {
    "metadata": {
      "date": "2017-01-01"
    },
    "feature": [
      "Feature 1"
    ]
  },
  "0.1.0": {
    "metadata": {
      "date": "2017-01-01"
    },
    "feature": [
      "Feature 1"
    ]
  }
}
`;

exports[`should return changes in JSON 1`] = `
{
  "2.1.0": {
    "metadata": {
      "date": "2017-01-05"
    },
    "feature": [
      "Feature 1",
      "Feature 2"
    ],
    "fix": [
      "Fix 1"
    ]
  },
  "2.0.1": {
    "metadata": {
      "date": "2017-01-04"
    },
    "fix": [
      "Fix 1"
    ]
  },
  "2.0.0": {
    "metadata": {
      "date": "2017-01-03"
    },
    "breaking": [
      "Breaking Change 1"
    ],
    "feature": [
      "Feature 1"
    ]
  },
  "1.0.0": {
    "metadata": {
      "date": "2017-01-02"
    },
    "feature": [
      "Feature 1"
    ]
  },
  "0.2.0": {
    "metadata": {
      "date": "2017-01-01"
    },
    "feature": [
      "Feature 1"
    ]
  },
  "0.1.0": {
    "metadata": {
      "date": "2017-01-01"
    },
    "feature": [
      "Feature 1"
    ]
  }
}
`;

exports[`should return changes in Yaml 1`] = `
2.1.0:
  metadata:
    date: '2017-01-05'
  feature:
    - Feature 1
    - Feature 2
  fix:
    - Fix 1
2.0.1:
  metadata:
    date: '2017-01-04'
  fix:
    - Fix 1
2.0.0:
  metadata:
    date: '2017-01-03'
  breaking:
    - Breaking Change 1
  feature:
    - Feature 1
1.0.0:
  metadata:
    date: '2017-01-02'
  feature:
    - Feature 1
0.2.0:
  metadata:
    date: '2017-01-01'
  feature:
    - Feature 1
0.1.0:
  metadata:
    date: '2017-01-01'
  feature:
    - Feature 1

`;

exports[`should return changes in markdown 1`] = `
## 2.1.0 (2017-01-05)
### Features
* Feature 1
* Feature 2
### Fixes
* Fix 1

## 2.0.1 (2017-01-04)
### Fixes
* Fix 1

## 2.0.0 (2017-01-03)
### BREAKING CHANGES
* Breaking Change 1
### Features
* Feature 1

## 1.0.0 (2017-01-02)
### Features
* Feature 1

## 0.2.0 (2017-01-01)
### Features
* Feature 1

## 0.1.0 (2017-01-01)
### Features
* Feature 1

`;

exports[`should return the latest version 1`] = `
2.1.0:
  metadata:
    date: '2017-01-05'
  feature:
    - Feature 1
    - Feature 2
  fix:
    - Fix 1

`;

exports[`should return unreleased changes 1`] = `
breaking:
  - Breaking Change 1
feature:
  - Feature 1
  - Feature 2
fix:
  - Fix 1
  - Fix 2

`;

exports[`should return unreleased changes from change files 1`] = `
breaking:
  - Breaking 1
  - Breaking 2
feature:
  - Feature 1
  - Feature 2
fix:
  - Fix 1
  - Fix 2

`;

exports[`should return unreleased changes in markdown 1`] = `
### BREAKING CHANGES
* Breaking Change 1
### Features
* Feature 1
* Feature 2
### Fixes
* Fix 1
* Fix 2

`;
